### PR TITLE
[MLIR][AMDGPU]Add refactoring for shared-mem optimization

### PR DESCRIPTION
**Pull request by erman-gurses**
Addressing the issues in this PR: https://github.com/llvm/llvm-project/pull/81550

### CODE REVIEW

**Comment by erman-gurses** [comment]
Hi @ftynse, let me know if I am missing something.

File: mlir/include/mlir/Dialect/AMDGPU/Transforms/Transforms.h [hunk 1]
```diff
@@ -49,7 +49,8 @@ namespace amdgpu {
 mlir::LogicalResult optimizeSharedMemoryReadsAndWrites(Operation *parentOp,
                                                        Value memrefValue);
 
-void optimizeSharedMemoryReadsAndWritesOp(mlir::func::FuncOp funcOp);
+std::optional<mlir::LogicalResult>
+optimizeSharedMemoryReadsAndWritesOp(mlir::func::FuncOp funcOp);
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/AMDGPU/Transforms/Transforms.h:53]: Nit: there's no need to prefix with `mlir::` in the mlir code base.

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp [hunk 2]
```diff
@@ -218,7 +218,8 @@ mlir::amdgpu::optimizeSharedMemoryReadsAndWrites(Operation *parentOp,
   return success();
 }
 
-void amdgpu::optimizeSharedMemoryReadsAndWritesOp(func::FuncOp funcOp) {
+std::optional<mlir::LogicalResult>
```
Comment (ftynse) [inline mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:221]: I don't see why the result is optional, nothing returns `nullopt` here.

**Review by ftynse — APPROVED** [review]

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp (hunk 2 above)
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:221]: There is one more `return;` statement above after checking `hasSharedMemoryAddressSpace()`. There is no failure in this case it just checks if the shared memory relation exists or not. So I added an optional for that.

**Pull request merged**